### ISSUE TITLE
The future of Analysis

### ISSUE THREAD

**Issue by rhelmot** [angr/angr#3353]
Tagging @fmagin.

Our current system of analyses being classes which are silently partially initialized is a huge hack. We should have something at least a little bit nicer.

My suggestion is that analyses can be _any callable_, and the versions available on project.analyses are partial applications which apply `project` and `kb` kwargs. This enables all the same hackability of the current system while enabling running analyses outside of the project.analyses hub for purposes of type annotation, or simply just preferring the less-magic paradigm.

**Comment by fmagin** [comment]
I think it would be nice if we can find some API that is comfortable for dynamic usage in IPython but _also_ type annotation compatible. Otherwise you can never be quite sure if your IDE is really giving you _all_ usages of some analysis, or if there are code paths that still use the dynamic version. But this might be fundamentally add odds, because any typed approach I can think of would at least require importing the Analysis Class, which is more annoying in a IPython CLI context.

Allowing "any callable" to be an analysis would also make it even harder to find every analysis in an IDE setup because you can't just search for all classes inheriting from `Analysis` anymore.
If there is some new dynamic way for scripting I would be strongly in favor of some CI linting rule that just doesn't allow the highly dynamic usage in code that that should be merged into the core repositories and enforces the typed version instead. Then everyone can use the dynamic version for prototyping, scripting, CTF challenges, etc, but everyone working with the actual codebase can use the type based IDE features that are IMO really required to work with a codebase of this size. 

The first idea that came to mind when annotating the current analysis hack is to have a base class/interface `Analysis` with a constructor that takes arguments like `project`, `kb,` etc, and an abstract method `analyze` that takes the arguments that would have been passed to `__init__` (like the specific function to analyze). But IIRC changing the signature of a method that is being overridden is something linters and IDEs complain about, so I guess this shouldn't be done.

**Comment by rhelmot** [comment]
Can you describe a situation where you want to find all analyses with an IDE?

**Comment by fmagin** [comment]
Simply to explore the codebase in general IIRC.  It might have been `ForwardAnalysis` specifically. What was way more important in my experience though was finding every location where an Analysis (like RDA) is used to get an idea of how it can be used. 

Tangential: If an "Analysis" can be _any callable_ this raises the question for me what it means to be an "Analysis"? Solely the fact that it is somehow available (via dynamic shenanigans) under `project.analyses.`?

**Comment by rhelmot** [comment]
Yeah - imo there's nothing special about analysis other than the bit of sugar we provide in the Analysis class with respect to resilience and the progress interface, which shouldn't really be necessary for membership on project.analyses. I think at some point the vision was to have tools in the class which make it easier to write distributed analyses, but that seems to have failed to materialize over the last nine years, and imo it would better belong in a different package.

And with that in mind, my opinion about there being no point in being able to find references to any analysis should be self-apparent :)

**Comment by fmagin** [comment]
That seems kinda limiting for future features.
I'd push for having a more _rigid_ interface and some extra methods/properties that an analysis is expected to provide. I am currently writing a few analyses for Ghidra, and their interface requires e.g. a function to check if you can even sensible analyze the binary, their Analysis class has a priority system so you can define the order and "stage" of the analysis.

That analysis order is something I am annoyed with our internal analyses (which aren't angr analyses, but IMO should be): There is an implicit requirement tree between them, but at best there is some check at the start of the constructor, typically they just crash if some expected data was not available. Ghidra has some number based system with stages https://ghidra.re/ghidra_docs/api/ghidra/app/services/AnalysisPriority.html
But there could also be a system where you can explicitly specify another Analysis that has to run first, with an option to resolve this dependency tree automatically when calling an Analysis that has more complex requirements.
